fix: bias assignment for 3 players and less

maanasb01 opened the pull request:
This PR addresses issue #40. 

## Cause of the issue

- While starting the game, we are generating `room_communities`. In case of 3 or fewer players, we are randomly keeping 2 communities.
- Then, we are generating Player structs. In that, in the case of 3 players, we are assigning them each with one identity. For 2 players, we are randomly assigning 2 identities.
- Then, while assigning biases, "we are assigning them from `room_communities`", we are selecting all the room_communities except the current player's identity.

### Scenarios:
#### 2 Players:
- Randomly selected room_communities: red, blue
- While Player Creation
  - Randomly selected player identities: red, yellow
  - Selecting Biases: (room_communities) - (player's identity)
    - Red Player Biases: Blue
    - Yellow Player Biases: Red, Blue 
    
**BLUE player doesn't exist, but blue bias is still being assigned.** 

##### 3 Players:
  - Randomly selected room_communities: red, blue
  - While Player Creation
    - Randomly selected player identities: red, blue, yellow
    - Selecting Biases: (room_communities) - (player's identity)
      - Red Player Biases: Blue
      - Blue Player Biases: Red
      - Yellow Player Biases: Red, Blue
      
**There is a yellow player, but `room_communities` doesn't have a yellow community.** 

## Fixes
- For 3 players, instead of assigning 2 room communities, now all 3 communities are being assigned. 
- For 2 players, I have added a case while creating Player structs, in case of 2 players, assign them based on room_communities (that are already randomly selected) only, and not randomly take 2 out of 3.
### Test Cases
- As I have added a new parameter in the function that assigns identities to the players, there was an issue with a test case: https://github.com/tattle-made/viral-spiral/blob/447a7a2804acc32ef73ff4052b617d41646adc4b/test/viral_spiral/room/card_draw_test.exs#L197-L230
- This test case checks for 3-12 players. As the newly added parameter is only used in the case of 2 players, I have passed `nil` as the second argument in the function. 
- I have created a new test case for 2 players.

dennyabrain commented on lib/viral_spiral/room/card_draw.ex:
```diff
@@ -102,6 +102,10 @@ defmodule ViralSpiral.Room.CardDraw do
         ^identity_count ->
```
@maanasb01 it took me a few re-reads to realize that what you are trying to handle here is the case when the number of players is equal to the number of total identities supported by viral spiral(3 here)

I wonder if the code was more readable if we created 2 conditions within this case.
one when `total` is 2 or 3, in which case just return `Enum.shuffle(room_communities)` and for all other cases (`_`) do whats already written.

Lmk if I am missing something.

dennyabrain replied:
Haan sure that works. My main issue was that it was hard to understand what scenario `^identity_count` was handling. With what you are suggesting, it would be clear that there is a default scenario and some special scenario for when the player count is 2.

dennyabrain commented on test/viral_spiral/room/card_draw_test.exs:
```diff
@@ -201,7 +201,9 @@ defmodule ViralSpiral.Room.CardDrawTest do
-        result = CardDraw.assign_player_identity(names)
+        # assign_player_identity function accepts room_communities as second argument that is only used in
+        # case of 2 players. Here, as the test is for 3 or more players, the second argument passed is nil
+        result = CardDraw.assign_player_identity(names, nil)
```
While I see how you've achieved what you want to with passing `nil` here, its a bit unclear what `room_communities` being nil means logically.

I understand that implicitly you have created a function that you want to call in different ways when the number of players is 2 vs >2, but the function definition itself doesn't make that clear. At the minimum, add documentation to the function and mention there the ways in which you want this `assign_player_identity` function to be called.

So if you must do it this way by passing `room_comunities` as nil, the idiomatic elixir way would be to change `assign_player_identity` definition to the following
```elixir
def assign_player_identity(names, room_communities \\ nil) do
```
This way you are communicating that this function can be called without passing `room_communities` parameter and when its not passed, by default its value would be nil.

Then in your test or any calling code, you can call this function in 2 ways
```elixir
CardDraw.assign_player_identity(["adhiraj", "aman", "farah", "krys"], [:red, :blue] )

CardDraw.assign_player_identity(["adhiraj", "aman])
```


The fact that you have 2 different test scenarios for when number of players is 2 and >=2 is a sign that some implementation detail is bleeding into the calling code. Ideally we should try to see if we can expand that same test to suport for all scenarios.

maanasb01 commented:
Hey @dennyabrain, I have pushed the changes. 
- For the `assign_player_identity` function, I have kept only 2 cases for when there are 2 players and when there are more than 2 players.
- For the test function, I have updated it to keep only one test case. I have updated the test case to work for 2-12 players (earlier it was 3-12). For 2 players, I am passing 2 communities, and for the rest, I am passing all the communities (which is the same as when the function would get executed with 3 or more players in real, while playing the game as well).

dennyabrain commented:
@maanasb01 Thank you. This is looking good. lets merge and mark for testing.
